software/contrib/cvecorder.py

```diff
@@ -0,0 +1,348 @@
+                cvs[i].voltage(self.CVR[self.ActiveBank][i][self.step])
+
+        # Reset step number at stepLength -1 as pattern arrays are zero-based
+        if self.step < self.stepLength - 1:
+            self.step += 1
+        else:
+            # Reset step to zero , stop recording and save recording to local storage
+            self.step = 0
+            if self.CvRecording[self.ActiveCvr] == 'true':
+                self.CvRecording[self.ActiveCvr] = 'false'
+                self.bankToSave = self.ActiveBank
+                self.saveState()
+
+    def clearCvrs(self, bank):
+        for b in range(self.numCVRBanks+1):
+            # skip bank unless 'all' is passed
+            if b != bank and bank != 'all':
+                continue
+            if self.debugTest:
+                print('Clearing bank: ' + str(b))
+            # Set all CV values to zero
+            for i in range(self.numCVR+1):
+                for n in range (0, self.stepLength):
+                    self.CVR[b][i][n] = 0
+            # Save the cleared bank to local storage
+            self.bankToSave = b
+            self.saveState()
+
+    def initCvrs(self):
+        for b in range(self.numCVRBanks+1):
+            self.CVR.append([])
+            for i in range(self.numCVR+1):
+                self.CVR[b].append([])
+                self.CvRecording.append('false')
+                for n in range (0, self.stepLength):
+                    self.CVR[b][i].append(0)
+
+    def saveState(self):
+        # generate output filename
+        outputFile = f"saved_state_{self.__class__.__qualname__}_{self.bankToSave}.txt"
+
+        # Convert each value to an int by multiplying by 100. This saves of storage and memory a little
+        i=0
+        for channel in self.CVR[self.bankToSave]:
+            self.CVR[self.bankToSave][i] = [int(x * 100) for x in self.CVR[self.bankToSave][i]]
+            i += 1
+
+        if self.debugTest:
+            print('Saving state for bank: ' + str(self.bankToSave))
+
+        # Trigger garbage collection to minimize memory use
+        gc.collect()
+
+        # Show free memory if running a debug test
+        if self.debugTest:
+            print(self.free())
+
+        # Write the value to a the state file
+        maxRetries = 6
+        attempts = 0
+        while attempts < maxRetries:
+            try:
+                attempts += 1
+                # Create json object of current CV bank
+                jsonState = json.dumps(self.CVR[self.bankToSave])
+                with open(outputFile, 'w') as file:
+                    # Attempt write data to state on disk, then break from while loop if the return (num bytes written) > 0
+                    if file.write(jsonState) > 0:
+                        file.close()
+                        break
+            except MemoryError as e:
+                if self.debugTest:
+                    print(f'[{attempts}] Error: Memory allocation failed, retrying: {e}')
```

Comment:
What changes on the subsequent attempts that allows the allocation to succeed?

Reply:
Nothing other than a small passage of time that may have allowed micropython to somehow (and rather hopefully!) fix itself. I do need to clean this section up a little as it is probably no longer needed now I have got to the bottom of the heap fragmentation issue. leave this with me!